### PR TITLE
(feat): obfuscate transactions and traces path params with config regex

### PR DESCRIPTION
**Pull request by matewilk**
Changes:

Based on the `url_obfuscation_regex` param from the `newrelic.js` config file.
- obfuscate `request.uri`
- obfuscate `http.url` (path part only)

For Transactions and Distributed Tracing (for incoming and outgoing http requests)

It also obfuscates `attributes` (with the same names as above) for Distributed Tracing (Spans and root of the transaction)

Adds Transaction Trace details `url` obfuscation.

### CODE REVIEW

File: lib/instrumentation/core/http-outbound.js [hunk 1]
```diff
@@ -205,16 +205,18 @@ function assignOutgoingHeaders(currentHeaders, outboundHeaders) {
...
  * Starts the http outbound segment and attaches relevant attributes to the segment/span.
  *
+ * @param {Agent} agent New Relic agent
  * @param {string|object} opts a url string or HTTP request options
  * @param {Function} makeRequest function to make request
  * @param {string} hostname host of outbound request
  * @param {TraceSegment} segment outbound http segment
  * @returns {http.IncomingMessage} request actual http outbound request
  */
-function applySegment(opts, makeRequest, hostname, segment) {
```
Comment (jordigh) [inline lib/instrumentation/core/http-outbound.js]: I believe `segment.transaction.agent` already has the agent.

Comment (matewilk) [reply lib/instrumentation/core/http-outbound.js]: Brilliant, thanks for the info, will check this out

File: test/unit/instrumentation/http/http.test.js [hunk 2]
```diff
@@ -687,7 +711,7 @@ test('built-in http module instrumentation', (t) => {
       const traceparent = '00-4bf92f3577b34da6a3ce929d0e0e4736-00f067aa0ba902b7-00'
       const priority = 0.789
       // eslint-disable-next-line
-        const tracestate = `190@nr=0-0-709288-8599547-f85f42fd82a4cf1d-164d3b4b0d09cb05-1-${priority}-1563574856827`
+        const tracestate = `190@nr=0-0-709288-8599547-f85f42fd82a4cf1d-164d3b4b0d09cb05-1-${priority}-1563574856827`;
```
Comment (bizob2828) [inline test/unit/instrumentation/http/http.test.js:719]: we don't use semi-colons and honestly I'm not sure how this isn't failing linting

File: lib/config/default.js [hunk 3]
```diff
@@ -1208,7 +1208,12 @@ defaultConfig.definition = () => ({
       formatter: boolean,
       default: false
     }
-  }
+  },
+
+  /**
+   * Default URL obfuscation regex - will never match anything
+   */
+  url_obfuscation_regex: '~^'
```
Comment (bizob2828) [inline lib/config/default.js]: we should set this to null. If we set it to a value that means we'll be running a regex eval on every url which is extra overhead we don't want our customers incurring

Comment (bizob2828) [reply lib/config/default.js]: might be good to put in the comment what this actually obfuscates. i know it says URL but it appears to do it in several places. also i know this is new and I don't see anything in our specs but not sure if this is the name we want to go with.

File: lib/util/urltils.js [hunk 4]
```diff
@@ -162,6 +162,23 @@ module.exports = {
...
   },
 
+  /**
+   * Obfuscates path parameters with regex from config
+   *
+   * @param {Config} config The configuration containing the regex
+   * @param {string} requestPath The Path to be obfuscated
+   * @returns {string} The obfuscated path
+   */
+  obfuscatePath: function obfuscatePath(config, requestPath) {
```
Comment (bizob2828) [inline lib/util/urltils.js]: i'd like to see a full suite of unit tests for this.  i know we didn't have any for the other urlutils before but should only need a few cases

File: lib/config/default.js (hunk 3 above)
Comment (bizob2828) [reply lib/config/default.js]: This is very misleading as it expects to be a regex but in the util you aren't creating a RegExp from a string, just assuming it's already a literal regex.  I'd keep this simple and parse the configuration value as a `new RegExp(config.url_obfuscation_regex)`. we don't have a regexp formatter but I would build one as we only want to do this once if the value is truthy

File: lib/util/urltils.js [hunk 5]
```diff
@@ -162,6 +162,23 @@ module.exports = {
...
+    if (typeof requestPath === 'string') {
+      const regex = config.url_obfuscation_regex
+      if (regex) {
+        return (requestPath = requestPath.replace(regex, '/***'))
```
Comment (bizob2828) [inline lib/util/urltils.js]: this feels very limiting.  I picture the user case of such:

`/api/v1/users/sensitive/edit`

there's currently no way for me to apply the url obfuscation regex to result in `/api/v1/users/**/edit`.  I'm leaning towards that configuration stanza being more open like this

```
url_obfuscation: {
  regex: '(/api/v1/users/)(.*)(/edit)',
  replacement: '$1**$3'
}
```

**Review by bizob2828 — CHANGES_REQUESTED** [review]
overall this looks like a good start.  have a few comments and thoughts. please let me know if you need clarification.  also how confident are you that you updated the write places to obfuscate the URLs? I'm not well versed in all the places that need updated.  also looking back through the slack thread it seems as though they are happy with a combo of attributes.exclude and this?

File: lib/config/default.js (hunk 3 above)
Comment (matewilk) [inline lib/config/default.js:1216]: test

**Comment by matewilk** [comment]
Thanks for the comprehensive review @bizob2828 

I'll need some guidance on two things:
- how the[ default.js](https://github.com/matewilk/node-newrelic/blob/main/lib/config/default.js) config file is merged with a `newrelic.js` file (coming from an app)
- when `formatters` are used and how they are being triggered on the app start.

I've made some changes on my local including a formatter:

in `default.js` I've included:
```
   /**
   * Default URL obfuscation regex - will never match anything
   */
  url_obfuscation: {
    formatter: regex, // formatter implementation doesn't matter for now
    default: null,
  },
```

And in the `newrelic.js` file:
```
  url_obfuscation: {
    regex: /.*/,
  },
```

I've put a bunch of breakpoints and tried figuring out how the two configs are merged, but I can't find the exact place. I know part of the logic is happening in[ this method](https://github.com/matewilk/node-newrelic/blob/main/lib/config/index.js#L969) but I'm struggling to identify the place where the merge happens.
For example for `attributes.exclude` this works as expected but with the above changes for `url_obfuscation` I ending up with `config.regex = /.*/` instead of `config.url_obfuscation.regex = /.*/`

I've also put breakpoints at [array](https://github.com/matewilk/node-newrelic/blob/main/lib/config/formatters.js#L16) and [boolean](https://github.com/matewilk/node-newrelic/blob/main/lib/config/formatters.js#L47) formatters but with the default `newrelic.js` config they where never fired.

**Pull request merged**